In [3]:
import json
import pandas as pd

In [4]:
df_data = pd.read_csv("hr_raw_data.csv", index_col=0)
data = df_data

In [5]:
data.columns

Index(['age', 'attrition', 'businesstravel', 'dailyrate', 'department',
       'distancefromhome', 'education', 'educationfield', 'employeecount',
       'employeenumber', 'environmentsatisfaction', 'gender', 'hourlyrate',
       'jobinvolvement', 'joblevel', 'jobrole', 'jobsatisfaction',
       'maritalstatus', 'monthlyincome', 'monthlyrate', 'numcompaniesworked',
       'over18', 'overtime', 'percentsalaryhike', 'performancerating',
       'relationshipsatisfaction', 'standardhours', 'stockoptionlevel',
       'totalworkingyears', 'trainingtimeslastyear', 'worklifebalance',
       'yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'roledepartament', 'numberchildren', 'remotework'],
      dtype='object')

In [5]:
columnas_deseadas = [
    'over18',
    'overtime',
    'percentsalaryhike',
    'performancerating',
    'relationshipsatisfaction',
    'standardhours',
    'stockoptionlevel',
    'totalworkingyears',
    'trainingtimeslastyear',
    'worklifebalance'
]

df_miscol = df_data[columnas_deseadas]

In [6]:
df_miscol

over18 overtime  percentsalaryhike performancerating  \
0         Y       No                 13               3,0   
1       NaN      NaN                 14               3,0   
2       NaN       No                 11               3,0   
3         Y      NaN                 19               3,0   
4         Y       No                 12               3,0   
...     ...      ...                ...               ...   
1673    NaN      NaN                 12               3,0   
1674      Y       No                 11               3,0   
1675    NaN       No                 19               3,0   
1676    NaN      Yes                 12               NaN   
1677      Y      Yes                 14               3,0   

      relationshipsatisfaction standardhours  stockoptionlevel  \
0                            3     Full Time                 0   
1                            1           NaN                 1   
2                            4           NaN                 0   
3                            2     Full Time                 2   
4                            4           NaN                 1   
...                        ...           ...               ...   
1673                         4           NaN                 0   
1674                         3     Full Time                 1   
1675                         1     Part Time                 0   
1676                         3     Part Time                 1   
1677                         4     Part Time                 0   

     totalworkingyears  trainingtimeslastyear worklifebalance  
0                  NaN                      5             3,0  
1                 34,0                      5             3,0  
2                 22,0                      3             NaN  
3                  NaN                      2             NaN  
4                  NaN                      5             3,0  
...                ...                    ...             ...  
1673               NaN                      2             3,0  
1674              27,0                      2             3,0  
1675               6,0                      1             3,0  
1676               NaN                      4             3,0  
1677              10,0                      2             3,0  

[1678 rows x 10 columns]

In [8]:
df_data.isnull()

age  attrition  businesstravel  dailyrate  department  \
0     False      False            True      False        True   
1     False      False            True      False        True   
2     False      False           False      False       False   
3     False      False           False      False        True   
4     False      False            True      False        True   
...     ...        ...             ...        ...         ...   
1673  False      False            True      False        True   
1674  False      False            True      False        True   
1675  False      False           False      False        True   
1676  False      False           False      False        True   
1677  False      False            True      False        True   

      distancefromhome  education  educationfield  employeecount  \
0                False      False            True          False   
1                False      False           False          False   
2                False      False           False          False   
3                False      False           False          False   
4                False      False           False          False   
...                ...        ...             ...            ...   
1673             False      False           False          False   
1674             False      False            True          False   
1675             False      False            True          False   
1676             False      False           False          False   
1677             False      False           False          False   

      employeenumber  ...  yearsatcompany  yearsincurrentrole  \
0              False  ...           False                True   
1              False  ...           False                True   
2              False  ...           False                True   
3              False  ...           False                True   
4              False  ...           False                True   
...              ...  ...             ...                 ...   
1673           False  ...           False                True   
1674           False  ...           False                True   
1675           False  ...           False                True   
1676           False  ...           False                True   
1677           False  ...           False                True   

      yearssincelastpromotion  yearswithcurrmanager  sameasmonthlyincome  \
0                       False                 False                False   
1                       False                 False                 True   
2                       False                 False                 True   
3                       False                 False                False   
4                       False                 False                False   
...                       ...                   ...                  ...   
1673                    False                 False                False   
1674                    False                 False                False   
1675                    False                 False                 True   
1676                    False                 False                False   
1677                    False                 False                 True   

      datebirth  salary  roledepartament  numberchildren  remotework  
0         False   False             True            True       False  
1         False   False             True            True       False  
2         False   False            False            True       False  
3         False   False             True            True       False  
4         False    True             True            True       False  
...         ...     ...              ...             ...         ...  
1673      False    True             True            True       False  
1674      False   False             True            True       False  
1675      False   False             True            True       False  
1676  

In [9]:
df_data.isnull().sum()

age                            0
attrition                      0
businesstravel               801
dailyrate                      0
department                  1366
distancefromhome               0
education                      0
educationfield               774
employeecount                  0
employeenumber                 0
environmentsatisfaction        0
gender                         0
hourlyrate                  1267
jobinvolvement                 0
joblevel                       0
jobrole                        0
jobsatisfaction                0
maritalstatus                675
monthlyincome                489
monthlyrate                    0
numcompaniesworked             0
over18                       938
overtime                     696
percentsalaryhike              0
performancerating            200
relationshipsatisfaction       0
standardhours                351
stockoptionlevel               0
totalworkingyears            549
trainingtimeslastyear          0
worklifeba

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1678 entries, 0 to 1677
Data columns (total 41 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       1678 non-null   object 
 1   attrition                 1678 non-null   object 
 2   businesstravel            877 non-null    object 
 3   dailyrate                 1678 non-null   float64
 4   department                312 non-null    object 
 5   distancefromhome          1678 non-null   int64  
 6   education                 1678 non-null   int64  
 7   educationfield            904 non-null    object 
 8   employeecount             1678 non-null   int64  
 9   employeenumber            1678 non-null   int64  
 10  environmentsatisfaction   1678 non-null   int64  
 11  gender                    1678 non-null   int64  
 12  hourlyrate                411 non-null    float64
 13  jobinvolvement            1678 non-null   int64  
 14  joblevel     

In [11]:
print(df_data['age'].unique())

['51' '52' '42' '47' '46' '48' '59' '41' '56' '38' '55' '40' '58' '35'
 '45' '33' '36' '34' 'forty-seven' '53' '43' '60' '32' '37' '49' '39' '50'
 '44' '30' 'fifty-eight' '29' '31' '54' '57' '27' 'thirty-six' '28' '26'
 'fifty-five' '25' 'fifty-two' 'thirty-one' '24' 'thirty' '23' '22' '21'
 '20' 'twenty-six' '19' 'thirty-seven' '18' 'thirty-two' 'twenty-four']


In [12]:
print(df_data['over18'].unique())

['Y' nan]


In [13]:
data['overtime'].unique()

array(['No', nan, 'Yes'], dtype=object)

In [14]:
data['percentsalaryhike'].unique()

array([13, 14, 11, 19, 12, 25, 16, 17, 22, 23, 20, 15, 21, 24, 18],
      dtype=int64)

In [15]:
data['performancerating'].unique()

array(['3,0', '4,0', nan], dtype=object)

In [16]:
data['relationshipsatisfaction'].unique()

array([3, 1, 4, 2], dtype=int64)

In [17]:
data['standardhours'].unique()

array(['Full Time', nan, 'Part Time'], dtype=object)

In [18]:
data['totalworkingyears'].unique()

array([nan, '34,0', '22,0', '28,0', '20,0', '21,0', '33,0', '40,0',
       '18,0', '25,0', '15,0', '17,0', '26,0', '16,0', '24,0', '14,0',
       '23,0', '27,0', '19,0', '11,0', '38,0', '37,0', '13,0', '12,0',
       '29,0', '10,0', '36,0', '35,0', '9,0', '31,0', '32,0', '8,0',
       '7,0', '30,0', '6,0', '5,0', '4,0', '3,0', '2,0', '1,0', '0,0'],
      dtype=object)

In [19]:
nan_indices = data[data["totalworkingyears"].isna()].index

Estudiando columna Overtime

In [20]:
NaN_OT = data[data["overtime"].isna()]

In [21]:
NaN_OT[["employeenumber", "jobrole", "department", "monthlyincome", "yearsatcompany", "overtime"]]

employeenumber                      jobrole                department  \
1                  2                     ManAGeR                        NaN   
3                  4           ReseArCH DIrECtOr                        NaN   
6                  7             Sales ExeCutIVe                        NaN   
17                18                     MAnaGER                        NaN   
18                19          REseArcH sciENtiST                        NaN   
...              ...                          ...                       ...   
1667            1268          rESEArcH SCiEntIst    Research & Development    
1669             977          rEseaRcH scIEnTiST                        NaN   
1670            1342          rEsEARCh ScieNTist                        NaN   
1671            1608   HEaltHCarE REPreSENtAtIve                        NaN   
1673             824          rESEaRcH SciEnTiST                        NaN   

     monthlyincome  yearsatcompany overtime  
1              NaN              33      NaN  
3        14307,50$              20      NaN  
6         8339,32$              21      NaN  
17       12783,92$              26      NaN  
18        4492,84$              16      NaN  
...            ...             ...      ...  
1667      4492,84$               1      NaN  
1669      2876,67$               3      NaN  
1670      1882,50$               0      NaN  
1671           NaN               9      NaN  
1673      3949,17$               3      NaN  

[696 rows x 6 columns]

In [22]:
NaN_OT["jobrole"].value_counts()
NaN_OT["department"].value_counts()

department
Research & Development     90
Sales                      42
Human Resources             4
Name: count, dtype: int64

In [23]:
data['overtime'].isna().sum()

696

In [24]:
data['performancerating'].value_counts()


performancerating
3,0    1257
4,0     221
Name: count, dtype: int64

In [12]:
#Over18 cambio NaN por Y pero se puede quitar esta columna:

df_miscol['over18'] = df_miscol['over18'].fillna('Y')

C:\Users\PC\AppData\Local\Temp\ipykernel_9804\96786770.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_miscol['over18'] = df_miscol['over18'].fillna('Y')


In [8]:
#cambiar la , por el .

df_miscol["worklifebalance"] = df_miscol["worklifebalance"].str.replace(",", ".")

C:\Users\PC\AppData\Local\Temp\ipykernel_9804\3389886655.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_miscol["worklifebalance"] = df_miscol["worklifebalance"].str.replace(",", ".")


In [9]:
df_miscol["totalworkingyears"] = df_miscol["totalworkingyears"].str.replace(",", ".")

C:\Users\PC\AppData\Local\Temp\ipykernel_9804\1630643654.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_miscol["totalworkingyears"] = df_miscol["totalworkingyears"].str.replace(",", ".")


In [10]:
df_miscol["performancerating"] = df_miscol["performancerating"].str.replace(",", ".")

C:\Users\PC\AppData\Local\Temp\ipykernel_9804\1075808933.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_miscol["performancerating"] = df_miscol["performancerating"].str.replace(",", ".")


In [29]:
data.groupby("performancerating")["percentsalaryhike"].mean().sort_values()

performancerating
3,0    13.978520
4,0    21.923077
Name: percentsalaryhike, dtype: float64

In [30]:
data[data['performancerating'].isna()].groupby("department").size()

department
Human Resources             2
Research & Development     37
Sales                       6
dtype: int64

In [31]:
data['over18']. unique()

array(['Y'], dtype=object)

CAMBIOS/Notas:

- Over18: Cambio NaN por Y (se puede quitar esta columna)

- Overtime. Valores 268 Yes, 714 No, 696 NaN. (Cambiar NaN por valor medio)

- percentsalaryhike: Valores (11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25) - Mucha desigualdad en incremento de salario. - No cambios necesarios.

- performancerating: 3,0 (1257), 4,0 (221), NaN (200). Cambiar los NaN por 3,0 o 4,0. - Averiguar escala. No se está haciendo una correcta evaluación del desempeño. (, cambiadas por .)

- relationshipsatisfaction: Valores (1 (318), 2 (350), 3 (527), 4 (483)). 318 empleados figuran en el número 1 - fomentar actividades/eventos de empresa - No cambios necesario 

- standardhours: Valores (Full Time' (400),  'Part Time' (927), NaN (351)). Averiguar 351 NaN (o calcular valor medio)

- stockoptionlevel: Valores (0 (719), 1 (689), 2 (177), 3 (93)) - No cambios necesario

- totalworkingyears: Valores (Floats, NaN) .- averiguar los NaN	549 NaN / Check si conservamos columna

- trainingtimeslastyear: Valores (0 (63), 1 (81), 2 (625), 3 (557), 4 (143), 5 (137), 6 (72)) - Aquí podemos averiguar que 63 empleados no recibieron ningún training en el último año, mientras que 72 recibieron 6.(Checkear relación por departamentos).-  No cambios necesario.

- worklifebalance: Valores (1.0 (82), 2.0 (374), 3.0 (946), 4.0 (162), NaN (114)) .- Averiguar los NaN	82 empleados figuran en el 1.0; la mayoria en el 3.0 .- (, cambiadas por .)

In [13]:
df_miscol.to_csv('columnas21-30.csv', index=False)

In [14]:
df_data['NUMCOMPANIESWORKED']. unique()

KeyError: 'NUMCOMPANIESWORKED'

In [10]:
df_miscol["totalworkingyears"] = df_miscol["totalworkingyears"].str.replace(",", ".")

C:\Users\PC\AppData\Local\Temp\ipykernel_15816\1630643654.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_miscol["totalworkingyears"] = df_miscol["totalworkingyears"].str.replace(",", ".")


--------------------

NULOS

In [ ]:
#OVER18 : Quitar o poner todos Y
#OVERTIME: Se podria calcular con el valor q mas tiene ('No');  (problema, tengo 600 NaN)
#Performancerating: Escala ordinal (1-4). Relleno con el valor más frecuente ('3.0')

# StandardHours: Uso de otras columnas (monthlyincome); MISSING 

#Totalworkingyears: Debe guardar relación con Yearsatthecompany y Numcompaniesworked -- SimpleImputer

#Worklifebalance: Debe guardar relación con environment y jobsatisfaction. Imputación por moda (valor más frecuente)

In [13]:
#Over18 cambio NaN por Y pero se puede quitar esta columna:

df_miscol['over18'] = df_miscol['over18'].fillna('Y')

#OVERTIME: Relleno con el valor más frecuente ('No')
df_miscol['overtime'] = df_miscol['overtime'].fillna('No')

#Performancerating: Escala ordinal (1-4). Relleno con el valor más frecuente ('3.0')

df_miscol['performancerating'] = df_miscol['performancerating'].fillna(3.0)

#Standardhours

#Totalworkingyears: Imputado por la media

from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy='mean')  # Imputar con la media
df_miscol['totalworkingyears'] = imputer.fit_transform(df_miscol[['totalworkingyears']])

#Worklifebalance

df_miscol['worklifebalance'].fillna(df_miscol['worklifebalance'].mode()[0], inplace=True)

C:\Users\PC\AppData\Local\Temp\ipykernel_15816\3603361865.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_miscol['over18'] = df_miscol['over18'].fillna('Y')
C:\Users\PC\AppData\Local\Temp\ipykernel_15816\3603361865.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_miscol['overtime'] = df_miscol['overtime'].fillna('No')
C:\Users\PC\AppData\Local\Temp\ipykernel_15816\3603361865.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_ind

In [14]:
df_miscol

over18 overtime  percentsalaryhike performancerating  \
0         Y       No                 13               3,0   
1         Y       No                 14               3,0   
2         Y       No                 11               3,0   
3         Y       No                 19               3,0   
4         Y       No                 12               3,0   
...     ...      ...                ...               ...   
1673      Y       No                 12               3,0   
1674      Y       No                 11               3,0   
1675      Y       No                 19               3,0   
1676      Y      Yes                 12               3.0   
1677      Y      Yes                 14               3,0   

      relationshipsatisfaction standardhours  stockoptionlevel  \
0                            3     Full Time                 0   
1                            1           NaN                 1   
2                            4           NaN                 0   
3                            2     Full Time                 2   
4                            4           NaN                 1   
...                        ...           ...               ...   
1673                         4           NaN                 0   
1674                         3     Full Time                 1   
1675                         1     Part Time                 0   
1676                         3     Part Time                 1   
1677                         4     Part Time                 0   

      totalworkingyears  trainingtimeslastyear worklifebalance  
0              11.35961                      5             3,0  
1              34.00000                      5             3,0  
2              22.00000                      3             3,0  
3              11.35961                      2             3,0  
4              11.35961                      5             3,0  
...                 ...                    ...             ...  
1673           11.35961                      2             3,0  
1674           27.00000                      2             3,0  
1675            6.00000                      1             3,0  
1676           11.35961                      4             3,0  
1677           10.00000                      2             3,0  

[1678 rows x 10 columns]